### ISSUE TITLE
Group-dependent lead/lag functionality in grouped_df

### ISSUE THREAD

**Issue by docendodiscimus** [hadley/dplyr#946]
I would like to ask for a feature request that would allow more flexible, group-dependent lead/lag functionality in grouped data. 

Here's an example of how it can be done using data.table's devel version 1.9.5:

``` r
library(data.table) # 1.9.5 devel
setDT(mtcars)
setkey(mtcars, cyl)
cols = c("carb", "gear", "hp"); 
mtcars[, (cols) := shift(.SD, carb[1],  NA, "lead"), .SDcols=cols, by = cyl]
```

Similar code doesn't currently work in dplyr_0.4.1:

``` r
data(mtcars)
library(dplyr) # 0.4.1
mtcars %>%
  arrange(cyl) %>%
  group_by(cyl) %>%
  mutate(gear = lead(gear, carb[1]))
Error: could not convert second argument to an integer. type=LANGSXP, length = 3
```

Same error when using mutate_each:

``` r
mtcars %>%
  arrange(cyl) %>%
  group_by(cyl) %>%
  mutate_each(funs(lead(., carb[1])), c(carb, gear, hp))
Error: could not convert second argument to an integer. type=LANGSXP, length = 3
```

It would be a nice additional feature for future dplyr versions. Thanks a lot for developing great packages!

**Comment by hadley** [comment]
Can you please explain in words what that's supposed to do?

**Comment by docendodiscimus** [comment]
Sure. 
It's supposed to shift a number of columns (in the example: c("carb", "gear", "hp")) "upwards" i.e. take leading values, but which leading value shall be taken depends on the value in another column (in the example I use the first entry in column "carb" in each group.
Since the first value of "carb" in each group (by "cyl") is different, the effect is that for cyl == 4, the columns are shifted 1 row (lead(., 1)); for cyl == 6 the values are shifted 6 rows (lead(., 6)) and for cyl == 8 the values are shifted 2 rows (lead(., 2)). I hope that makes it more clear.

**Comment by hadley** [comment]
Do you have a practical use case for this? It seems rather esoteric to me.

**Comment by docendodiscimus** [comment]
Yes, in a project I'm working on. We were computing manufacturing lot sizes, initially without lead times and now we got additional data about article-specific lead times. Since this means that manufacturing needs to start x number of days earlier, I wanted to shift some columns "upwards" and how much depends on each article's production lead time.
I realize that it's not a very common use case, but I happened to encounter it today.
